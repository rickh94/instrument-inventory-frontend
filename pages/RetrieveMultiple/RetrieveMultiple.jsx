import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  Paper,
  withStyles,
  Typography,
  FormControl,
  InputLabel,
  Input,
  Button,
  InputAdornment,
  IconButton,
  ListItem,
  List,
  ListItemText,
  ListItemSecondaryAction,
  FormGroup,
  Dialog,
  DialogContent,
  DialogContentText,
  DialogActions,
  DialogTitle
} from '@material-ui/core'

import Scanner from '../../components/Scanner'

import DeleteIcon from '@material-ui/icons/Delete'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBarcode } from '@fortawesome/free-solid-svg-icons'
import { API } from 'aws-amplify'

const styles = {
  root: {
    padding: '1rem',
    margin: '1rem auto',
    maxWidth: 700
  },
  lastButton: {
    marginLeft: 'auto'
  },
  buttons: {
    width: '100%',
    display: 'flex'
  },
  centerButtons: {
    width: '100%',
    display: 'flex',
    justifyContent: 'space-around'
  }
}

class RetrieveMultiple extends Component {
  constructor(props) {
    super(props)

    this.state = {
      instrumentList: [],
      scanning: false,
      error: null,
      instrumentNumber: '',
      message: null
    }
  }

  onDetected = result => {
    const item = result.codeResult.code
    if (!this.state.instrumentList.includes(item)) {
      this.addToList(item)
    }
  }

  addToListFromForm = e => {
    e.preventDefault()
    this.addToList(this.state.instrumentNumber)
    this.setState({ instrumentNumber: '' })
  }

  addToList = item => {
    this.setState({
      instrumentList: this.state.instrumentList.concat([item.toUpperCase()])
    })
  }

  removeInstrumentIndex = idx => {
    let instrumentListCopy = this.state.instrumentList.slice(
      0,
      this.state.instrumentList.length
    )
    instrumentListCopy.splice(idx, 1)
    this.setState({ instrumentList: instrumentListCopy })
  }

  handleSubmit = async e => {
    e.preventDefault()
    this.setState({ error: 'Please provide an instrument' })

    try {
      const res = await API.post('instrument-inventory', 'retrieve-multiple', {
        body: { instrumentNumbers: this.state.instrumentList }
      })
      console.log(res)
    } catch (err) {
      console.error(err)
    }
  }

  render() {
    const { classes } = this.props
    return (
      <React.Fragment>
        <Paper className={classes.root}>
          <Typography variant="h5">Retrieve Multiple Instruments</Typography>
          {this.state.scanning ? (
            <div>
              <Scanner onDetected={this.onDetected} />{' '}
              <Button onClick={() => this.setState({ scanning: false })}>
                Stop Scanning
              </Button>
            </div>
          ) : (
            <form onSubmit={this.addToListFromForm} style={{ width: '100%' }}>
              <FormControl fullWidth error={this.state.error ? true : false}>
                <InputLabel htmlFor="instrument-number">Instrument Number</InputLabel>
                <Input
                  id="instrument-number"
                  onChange={e => this.setState({ instrumentNumber: e.target.value })}
                  aria-describedby="instrument-number-error"
                  type="text"
                  value={this.state.instrumentNumber}
                  endAdornment={
                    <InputAdornment position="end">
                      <IconButton
                        onClick={() =>
                          this.setState({ scanning: !this.state.scanning })
                        }
                      >
                        <FontAwesomeIcon icon={faBarcode} />
                      </IconButton>
                    </InputAdornment>
                  }
                />
                {this.state.error && (
                  <FormHelperText id="instrument-number-error">
                    {this.state.error}
                  </FormHelperText>
                )}
              </FormControl>
              <div className={classes.buttons}>
                <Button onClick={this.addToListFromForm} className={classes.lastButton}>
                  Add
                </Button>
              </div>
            </form>
          )}
          <Typography variant="h6" align="center">
            Instruments to Retrieve
          </Typography>
          <List>
            {this.state.instrumentList.map((item, idx) => (
              <ListItem key={idx}>
                <ListItemText>{item}</ListItemText>
                <ListItemSecondaryAction>
                  <IconButton
                    aria-label="Delete"
                    onClick={() => this.removeInstrumentIndex(idx)}
                  >
                    <DeleteIcon />
                  </IconButton>
                </ListItemSecondaryAction>
              </ListItem>
            ))}
          </List>
          <FormGroup row className={classes.centerButtons}>
            <Button onClick={() => this.setState({ instrumentList: [] })}>
              Clear All
            </Button>
            <Button onClick={this.handleSubmit} color="primary">
              Submit
            </Button>
          </FormGroup>
        </Paper>
        <Dialog
          open={this.state.error ? true : false}
          onClose={() => this.setState({ error: null })}
        >
          <DialogContent>
            <DialogContentText>{this.state.error}</DialogContentText>
            <DialogActions>
              <Button onClick={() => this.setState({ error: null })}>OK</Button>
            </DialogActions>
          </DialogContent>
        </Dialog>
        <Dialog
          open={this.state.message ? true : false}
          onClose={() => this.setState({ message: null })}
        >
          <DialogTitle>Success</DialogTitle>
          <DialogContent>
            <DialogContentText>{this.state.message}</DialogContentText>
          </DialogContent>
        </Dialog>
      </React.Fragment>
    )
  }
}

RetrieveMultiple.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(RetrieveMultiple)
