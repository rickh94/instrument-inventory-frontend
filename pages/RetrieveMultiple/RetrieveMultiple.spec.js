import React from 'react';
import RetrieveMultiple from './RetrieveMultiple';

describe('<RetrieveMultiple />', () => {
});
