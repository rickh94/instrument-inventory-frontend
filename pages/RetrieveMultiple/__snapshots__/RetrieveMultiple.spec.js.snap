// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RetrieveMultiple /> matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded WithStyles(ForwardRef(Paper))-root-4"
  >
    <div
      class="makeStyles-heading-32"
      data-testid="loading-header"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
      >
        Retrieve Multiple Instruments
      </h5>
    </div>
    <form
      style="width: 100%;"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="instrument-number"
        >
          Instrument Number
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
            id="instrument-number"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
          >
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-barcode fa-w-16 "
                  data-icon="barcode"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 448V64h18v384H0zm26.857-.273V64H36v383.727h-9.143zm27.143 0V64h8.857v383.727H54zm44.857 0V64h8.857v383.727h-8.857zm36 0V64h17.714v383.727h-17.714zm44.857 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm35.715 0V64h18v383.727h-18zm44.857 0V64h18v383.727h-18zm35.999 0V64h18.001v383.727h-18.001zm36.001 0V64h18.001v383.727h-18.001zm26.857 0V64h18v383.727h-18zm45.143 0V64h26.857v383.727h-26.857zm35.714 0V64h9.143v383.727H476zm18 .273V64h18v384h-18z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="RetrieveMultiple-buttons-2"
      >
        <button
          class="MuiButtonBase-root MuiButton-root RetrieveMultiple-lastButton-1 MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter"
    >
      Instruments to Retrieve
    </h6>
    <ul
      class="MuiList-root MuiList-padding"
    />
    <div
      class="MuiFormGroup-root MuiFormGroup-row RetrieveMultiple-centerButtons-3"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Clear All
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-textPrimary MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Submit
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
